### ISSUE TITLE
Question Regarding the Loss Design

### ISSUE THREAD

**Issue by ooochen-30** [Hui-design/TSPO#8]
First and foremost, thank you for your incredible work on TSPO and for making the code publicly available. It's a fantastic resource for the community, and I'm learning a great deal by studying it.

I'm currently digging into the implementation of the `LLaVAVideoTSPOTrainer` and have a question regarding the specific way the policy gradient loss is calculated in the compute_loss method. I would be very grateful if you could provide some clarification.
``` pyhton
# tspo_traner.py at line 594
loss_list, mean_kl_list, completion_length_list = 0.0, 0.0, 0
        for batch_idx in range(self.num_generations):
            # token_probs_item = torch.exp(per_token_logps_batch[batch_idx] - per_token_logps_batch[batch_idx].detach()).mean()
            ts_probs_item = torch.exp(ts_logps_batch[batch_idx] - ts_logps_batch[batch_idx].detach()).mean()
            # probs_loss = token_probs_item * ts_probs_item * advantages[batch_idx]
            probs_loss = ts_probs_item * advantages[batch_idx]
            # mean_kl = per_token_kl_batch[batch_idx].mean() + ts_kl_batch[batch_idx].mean()
            mean_kl = per_token_kl_batch[batch_idx].mean()
            # loss = - (probs_loss - self.beta * mean_kl)  
            loss = - (probs_loss) 
            loss_list += loss
            mean_kl_list += mean_kl 
            completion_length_list += all_completion_ids[batch_idx].shape[1]
```
1. My understanding is that `torch.exp(log_p - log_p.detach())` would mathematically evaluate to a tensor of ones, but it cleverly preserves the gradient path through the first log_p term,it seems to achieve a similar result with  REINFORCE loss. Could you please elaborate on the intuition behind choosing this specific formulation.
2.  The key KL divergence term (which is central to PPO/GRPO for constraining policy updates) appears to be omitted from the final loss calculation. Should I understand this implementation as a simplified version of GRPO, or is it more aligned with a REINFORCE-with-baseline style algorithm?

I apologize if my questions are based on a misunderstanding of the code. Any clarification on these points would be immensely helpful for my research.

Thank you again for your time and for the excellent work!

**Comment by ooochen-30** [comment]
As I was tracing the data flow for the loss calculation, I noticed one more small detail that I'd like to ask about, if you don't mind.

It seems there might be a return value mismatch in the LlavaQwenForCausalLM.temporal_sampling method. The function is defined to return two values (sel_idx, logp_ts) when method is None, but the call in LLaVAVideoTSPOTrainer unpacks three values (_, ts_logps_raw, pred_scores).
```python
# llava_qwen.py
    def temporal_sampling(self, image_embeddings, text_features, clip_scores=None, sample_len=64, ts_ids=None, 
                          window_size=None, score_tau=0.025, method=None):
        confidence, _ = self.get_confidence_score()(image_embeddings, text_features, clip_scores, window_size, score_tau)
        assert confidence.ndim == 1
        if method is None:
            if ts_ids is None:
                sel_idx, probs, logp_ts = gumbel_softmax(confidence.unsqueeze(1), sample_len=sample_len)
                sel_idx = (sel_idx.clone(), sel_idx)
            else: 
                _, probs, logp_ts = gumbel_softmax(confidence.unsqueeze(1), sample_len=sample_len)
                sel_idx = ts_ids
            return sel_idx, logp_ts, confidence
        else:
            return self.inference_ts(confidence, sample_len=sample_len, method=method)
```
I'm guessing the correct implementation should be return sel_idx, logp_ts, confidence to match the call. Could you please confirm this?

**Comment by Hui-design** [comment]
Thank you for the attention! 我非常乐意为您解答，我们欢迎你这样善于思考和提问的人：
